File: algo/lc.198.cc
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.empty()) return 0;
        
        vector<int> dp (nums.size(), 0);
        for (int i = 0; i < nums.size(); i++) {
            if (i==0) {
                dp[i] = nums[0];
            } else if (i==1) {
                dp[i] = nums[0]>nums[1] ? nums[0] : nums[1];
            } else {
                dp[i] = max(nums[i] + dp[i-2], dp[i-1]);
            }
        }
        return dp[nums.size()-1];
    }
};

File: algo/lc.198.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if not nums:
            return 0
        elif len(nums) == 1:
            return nums[0]
        else:
            # length >= 2
            dp = [nums[0], max(nums[1], nums[0])]
            for i in range(2, len(nums)):
                dpi = max(dp[i-1], dp[i-2] + nums[i])
                dp.append(dpi)
            return dp[-1]
